syne_tune/optimizer/schedulers/hyperband_stopping.py

```diff
@@ -34,58 +34,154 @@ def quantile_cutoff(values, prom_quant, mode):
     return np.quantile(values, q)
 
 
-class StoppingRungSystem(object):
+class RungSystem(object):
     """
-    Implements stopping rule resembling the median rule. Once a config is
-    stopped, it cannot be promoted later on.
-    This is different to what has been published as ASHA (see
-    :class:`PromotionRungSystem`).
+    Terminology: trials emit results at certain resource levels (e.g.,
+    epoch numbers). Some resource levels are rung levels, this is where
+    scheduling decisions (stop, continue or pause, resume) are taken.
+
+    For a running trial, the next rung level it will reach is called
+    its milestone.
+
     """
-    def __init__(
-            self, rung_levels, promote_quantiles, metric, mode, resource_attr):
-        # The data entry in _rungs is a dict mapping task_key to
-        # reward_value
-        assert len(rung_levels) == len(promote_quantiles)
-        self._mode = mode
+    def __init__(self, metric: str, mode: str, resource_attr: str):
         self._metric = metric
+        self._mode = mode
         self._resource_attr = resource_attr
-        self._rungs = [
-            RungEntry(level=x, prom_quant=y, data=dict())
-            for x, y in reversed(list(zip(rung_levels, promote_quantiles)))]
 
-    def on_task_schedule(self):
+    def on_task_schedule(self) -> dict:
+        """
+        Called when new task is to be scheduled. For a promotion-bqsed
+        rung system, check whether any trial can be promoted. If so,
+        return `dict(trial_id, resume_from, milestone)`, where
+        `resume_from` is the rung level where the trial (to be promoted)
+        is paused, and `milestone` is the next rung level it will reach
+        (may be None).
+
+        :return: dict(trial_id, resume_from, milestone) or dict()
+        """
         return dict()
 
-    def on_task_add(self, trial_id, skip_rungs, **kwargs):
+    def on_task_add(self, trial_id: str, skip_rungs: int, **kwargs):
+        """
+        Called when new task is started.
+
+        :param trial_id:
+        :param skip_rungs: This number of lowest rung levels are not
```

Comment:
It's a bit a odd that this is described here but this isn't happening here. Is it possible to move the corresponding logic into this function?

Reply:
Don't understand. These methods are more like callbacks, they are called when something happens

Reply:
The rationale of skip_rungs comes into play when you do Hyperband, where (a) each bracket has its own rung level system, or (b) all brackets share one system, but have different values of skip_rungs. So I cannot store skip_rungs in the rung level system, because in case (a) I don't need it (skip_rungs=0), and in case (b) I only have one system, but different values of skip_rungs.

I also don't understand what the problem is. These methods are just like callbacks, they are called when something happens. When I implement a class for callbacks, I don't have to implement all methods. And even if I do, I may not need all arguments passed into the method.

Reply:
I agree. Maybe one resolve this by removing skip_rungs here? What's strange to me is that the abstract class has skip_rungs as a param which isn't needed by all its children. So one could remove here and have it covered by kwargs and only introduce it in PromotionRungSystem. I won't insist it though.